pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java

```diff
@@ -673,7 +679,15 @@ public boolean isBackwardCompatibleWith(Schema oldSchema) {
       FieldSpec oldSchemaFieldSpec = entry.getValue();
       FieldSpec fieldSpec = getFieldSpecFor(oldSchemaColumnName);
       if (!fieldSpec.equals(oldSchemaFieldSpec)) {
-        return false;
+        if (fieldSpec.getDataType() == DataType.BOOLEAN) {
+          // For BOOLEAN data type, replace with STRING and compare again to keep backward-compatibility
+          fieldSpec.setDataType(DataType.STRING);
```

Comment:
It looks like if the schema is mutated here, we'll persist this updated schema to ZK. Would this schema work with data of new Boolean type? e.g. an ingestion job may be about to output some Boolean values to the table (not sure if this is a valid scenario).

Reply:
This is the desired behavior for existing table in the cluster. Changing `STRING` to `BOOLEAN` is illegal and can cause problem internally, so we want to keep the old behavior for the existing tables. For the new added tables, there is no such problem because the column is stored as `BOOLEAN` in the beginning.